File: DSIP Lab/Experiment-2/Linear_Convolution.py
"""
============================================================================
EXPERIMENT - 2: LINEAR CONVOLUTION
============================================================================

Author: Amey Thakur
GitHub: https://github.com/Amey-Thakur
Repository: https://github.com/Amey-Thakur/DIGITAL-SIGNAL-AND-IMAGE-PROCESSING-AND-DIGITAL-SIGNAL-AND-IMAGE-PROCESSING-LAB

Course: Digital Signal and Image Processing Lab (CSL701)
Roll No: 50
Batch: B3
Date: 30/07/2021

Description:
This program demonstrates Linear Convolution of two discrete-time signals
using Python. Linear convolution is a fundamental operation in Digital Signal
Processing used to determine the output of a Linear Time-Invariant (LTI) system
given its input and impulse response: y[n] = x[n] * h[n].

============================================================================
"""

# Import required libraries
import numpy as np              # For numerical operations
import matplotlib.pyplot as plt # For visualization

# ============================================================================
# INPUT SIGNAL DEFINITION
# ============================================================================

# Define input sequence x[n]
# This represents the input signal to the LTI system
Xn = np.array([1, 3, 2, -1])

# Define impulse response h[n]
# This represents the system's impulse response
Yn = np.array([1, -1, 3])

# ============================================================================
# LINEAR CONVOLUTION COMPUTATION
# ============================================================================

# Perform linear convolution using NumPy's convolve() function
# The output length will be: L = length(Xn) + length(Yn) - 1 = 4 + 3 - 1 = 6
C = np.convolve(Xn, Yn)

# Generate sample index vector for plotting
n = np.arange(1, len(C) + 1)

# ============================================================================
# VISUALIZATION: LINEAR CONVOLUTION RESULT
# ============================================================================

# Create figure
plt.figure(figsize=(10, 6))

# Display the convolution result using stem plot (discrete representation)
plt.stem(n, C, basefmt='k-')

# Add descriptive labels
plt.xlabel('n (Sample Index)')
plt.ylabel('y[n] (Amplitude)')
plt.title('Linear Convolution Result: y[n] = x[n] * h[n]')
plt.grid(True)

# Display the plot
plt.show()

# Print the convolution result
print("Input signal x[n]:", Xn)
print("Impulse response h[n]:", Yn)
print("Linear Convolution output y[n]:", C)

# ============================================================================
# END OF PROGRAM
# ============================================================================

File: DSIP Lab/Experiment-2/Linear_Convolution.m
% == == == == == == == == == == == == == == == == == == == == == == == == == ==
    == == == == == == == ==
    = % DIGITAL SIGNAL AND IMAGE PROCESSING LAB % % Experiment No. : 2 %
              Experiment Title
    : Linear and Circular Convolution of Discrete Signals %
      % Aim : % Write a program to perform Discrete Convolution such as %
              Linear Convolution and Circular Convolution.% %
              Student Details : % Name : Amey Thakur % Roll No. : 50 %
                                         Class : COMPS -
          BE - B -
          50 % Batch : B3 % % Date of Experiment : 30 / 07 / 2021 %
                       Date of Submission : 30 / 07 / 2021 % %
                       Author : % Amey Thakur % GitHub Profile : %
                                                                 https
    : // github.com/Amey-Thakur
      %
      % Repository : %
                     https
    : // github.com/Amey-Thakur/DIGITAL-SIGNAL-AND-IMAGE-PROCESSING-AND-DIGITAL-SIGNAL-AND-IMAGE-PROCESSING-LAB
      %
      == == == == == == == == == == == == == == == == == == == == == == == == ==
      == == == == == == == == ==
    =

        % Clear workspace,
    close all figures, and clear command window clear all;
close all;
clc;

% -- -- -- -- -- -- -- -- -- -- -- -- -- -- -% Input Discrete -
    Time Signals % -- -- -- -- -- -- -- -- -- -- -- -- -- -- -Xn = [1 3 2 - 1];
% First input sequence Yn = [1 - 1 3];
% Second input sequence

    % -- -- -- -- -- -- -- -- -- -- -- -- -- -- -% Linear Convolution %
    -- -- -- -- -- -- -- -- -- -- -- -- -- -- -C = conv(Xn, Yn);     % Perform linear convolution

% Sample index for output signal
n = 1:length(C);

% -- -- -- -- -- -- -- -- -- -- -- -- -- -- -% Plot the Result %
    -- -- -- -- -- -- -- -- -- -- -- -- -- -- -figure;
stem(n, C, 'filled');
grid on;
title('Linear Convolution of Two Discrete Signals');
xlabel('Sample Index');
ylabel('Amplitude');
